ext/mbstring/mbstring.c

```diff
@@ -3993,29 +3993,15 @@ static inline zend_long php_mb_ord(const char *str, size_t str_len, zend_string
 		return -2;
 	}
 
-	{
-		mbfl_wchar_device dev;
-		mbfl_convert_filter *filter;
-		zend_long cp;
-
-		mbfl_wchar_device_init(&dev);
-		filter = mbfl_convert_filter_new(enc, &mbfl_encoding_wchar, mbfl_wchar_device_output, 0, &dev);
-		/* If this assertion fails this means some memory allocation failure which is a bug */
-		ZEND_ASSERT(filter != NULL);
-
-		mbfl_convert_filter_feed_string(filter, (unsigned char*)str, str_len);
-		mbfl_convert_filter_flush(filter);
-
-		if (dev.pos < 1 || filter->num_illegalchar || dev.buffer[0] == MBFL_BAD_INPUT) {
-			cp = -1;
-		} else {
-			cp = dev.buffer[0];
-		}
+	uint32_t wchar_buf[6];
```

Comment:
Why 6? Is there some implicit contract on the minimum size of the buffer?

Reply:
Some of the existing fast conversion filters require the wchar buffer to have some minimum size, yes.

For an example, consider something like Base64. The natural way to decode Base64 is to process the input bytes in groups of 4, and to emit 3 wchars for each such group.

It would result in a lot of contortions if the conversion filter had to emit just 1 or 2 of these wchars and then stash the other one somewhere. So for that reason, the Base64 conversion filter requires the wchar buffer to have at least 3 spaces. If the buffer's size is not an even multiple of 3, the conversion filter will return with some empty spaces still remaining in the buffer. Remember the conversion filters return `out_len`, indicating how many wchars they produced, meaning it is totally legal for a conversion filter to return while not having used all the space in the wchar buffer (waiting until the next call to continue where it left off).

Other conversion filters have similar considerations; for some legacy encodings, the conversion code is greatly simplified if we can assume that there is some minimum size to the wchar buffer. In some of the filters I have used code like `ZEND_ASSERT(bufsize >= n)` so that if we ever violate that rule somewhere, we will find out about it.

I don't remember right now which of the legacy encodings needed the biggest minimum wchar buffer size, but I can check that.

Reply:
Hmm, so it looks like the filter which has the highest size requirement for the wchar buffer is UTF-7. It asserts that the buffer size is at least 5, but from reading the code, I think I could have made that 4 and everything would still work.

Reply:
Will look at this a bit more tomorrow.

Reply:
Regardless of the actual value, it likely makes sense to either introduce a self-describing macro, or to add a short comment about that value.

Reply:
@cmb69 Right you are.

So, it turns out that the legacy encoding which requires the greatest minimum buffer size is SJIS-Mac. Believe it or not, SJIS-Mac can produce an *incredible* 5 Unicode codepoints from just *one* kuten code! :scream: Crazy stuff.

Please see the update to this PR.

Reply:
OK, libfuzzer has reminded me of why I had set the min buffer size for UTF-7 and UTF7-IMAP to 5. I have added a comment to make it (hopefully) clearer.

libfuzzer agrees with me that the Base64 conversion filter only needs a min buffer size of 3.